**Стогний Е.А., Барсегян К.Д., М22-Ш02**

Импортируем необходимые библиотеки

In [1]:
import pm4py
from pm4py.objects.log.importer.xes import importer as xes_importer

Загрузим наши данные

In [2]:
log = xes_importer.apply('BPI_Challenge_2019.xes')
log

C:\Users\barse\AppData\Roaming\Python\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
parsing log, completed traces :: 100%|██████████| 251734/251734 [00:42<00:00, 5907.79it/s]


[{'attributes': {'Spend area text': 'CAPEX & SOCS', 'Company': 'companyID_0000', 'Document Type': 'EC Purchase order', 'Sub spend area text': 'Facility Management', 'Purchasing Document': '2000000000', 'Purch. Doc. Category name': 'Purchase order', 'Vendor': 'vendorID_0000', 'Item Type': 'Standard', 'Item Category': '3-way match, invoice before GR', 'Spend classification text': 'NPR', 'Source': 'sourceSystemID_0000', 'Name': 'vendor_0000', 'GR-Based Inv. Verif.': False, 'Item': '00001', 'concept:name': '2000000000_00001', 'Goods Receipt': True}, 'events': [{'User': 'batch_00', 'org:resource': 'batch_00', 'concept:name': 'SRM: Created', 'Cumulative net worth (EUR)': 298.0, 'time:timestamp': datetime.datetime(2018, 1, 2, 12, 53, tzinfo=datetime.timezone.utc)}, '..', {'User': 'user_002', 'org:resource': 'user_002', 'concept:name': 'Clear Invoice', 'Cumulative net worth (EUR)': 298.0, 'time:timestamp': datetime.datetime(2018, 3, 29, 13, 6, tzinfo=datetime.timezone.utc)}]}, '....', {'attrib

Основой для выполнения лабораторной работы выступает более полутора миллиона данных от крупной многонациональной компании, работающей в Нидерландах в области покрытий и красок. Рассматривается процесс обработки заказов на поставку для некоторых из ее 60 дочерних компаний. Журнал содержит данные с 1948 по 2018 гг.

Преобразуем данные в датафрейм

In [3]:
log_csv = pm4py.convert_to_dataframe(log)
log_csv

C:\Users\barse\AppData\Roaming\Python\Python311\site-packages\pm4py\utils.py:486: UserWarning: the EventLog class has been deprecated and will be removed in a future release.
  warnings.warn("the EventLog class has been deprecated and will be removed in a future release.")


User org:resource                         concept:name  \
0        batch_00     batch_00                         SRM: Created   
1        batch_00     batch_00                        SRM: Complete   
2        batch_00     batch_00               SRM: Awaiting Approval   
3        batch_00     batch_00              SRM: Document Completed   
4        batch_00     batch_00  SRM: In Transfer to Execution Syst.   
...           ...          ...                                  ...   
1595918  user_603     user_603   Change Approval for Purchase Order   
1595919  user_602     user_602           Create Purchase Order Item   
1595920  user_603     user_603   Change Approval for Purchase Order   
1595921  user_602     user_602           Create Purchase Order Item   
1595922  user_603     user_603   Change Approval for Purchase Order   

         Cumulative net worth (EUR)            time:timestamp  \
0                             298.0 2018-01-02 12:53:00+00:00   
1                             298.0 2018-01-02 13:53:00+00:00   
2                             298.0 2018-01-02 13:53:00+00:00   
3                             298.0 2018-01-02 13:53:00+00:00   
4                             298.0 2018-01-02 13:53:00+00:00   
...                             ...                       ...   
1595918                      1385.0 2019-01-17 14:00:00+00:00   
1595919                      1385.0 2019-01-17 13:32:00+00:00   
1595920                      1385.0 2019-01-17 14:00:00+00:00   
1595921                      1385.0 2019-01-17 13:32:00+00:00   
1595922                      1385.0 2019-01-17 14:00:00+00:00   

        case:Spend area text    case:Company case:Document Type  \
0               CAPEX & SOCS  companyID_0000  EC Purchase order   
1               CAPEX & SOCS  companyID_0000  EC Purchase order   
2               CAPEX & SOCS  companyID_0000  EC Purchase order   
3               CAPEX & SOCS  companyID_0000  EC Purchase order   
4               CAPEX & SOCS  companyID_0000  EC Purchase order   
...                      ...             ...                ...   
1595918                       companyID_0003    Framework order   
1595919                       companyID_0003    Framework order   
1595920                       companyID_0003    Framework order   
1595921                       companyID_0003    Framework order   
1595922                       companyID_0003    Framework order   

        case:Sub spend area text case:Purchasing Document  ...    case:Vendor  \
0            Facility Management               2000000000  ...  vendorID_0000   
1            Facility Management               2000000000  ...  vendorID_0000   
2            Facility Management               2000000000  ...  vendorID_0000   
3            Facility Management               2000000000  ...  vendorID_0000   
4            Facility Management               2000000000  ...  vendorID_0000   
...                          ...                      ...  ...            ...   
1595918                                        4508076348  ...  vendorID_1974   
1595919                                        4508076348  ...  vendorID_1974   
1595920                                        4508076348  ...  vendorID_1974   
1595921                                        4508076348  ...  vendorID_1974   
1595922                                        4508076348  ...  vendorID_1974   

        case:Item Type              case:Item Category  \
0             Standard  3-way match, invoice before GR   
1             Standard  3-way match, invoice before GR   
2             Standard  3-way match, invoice before GR   
3             Standard  3-way match, invoice before GR   
4             Standard  3-way match, invoice before GR   
...                ...                             ...   
1595918          Limit                     2-way match   
1595919          Limit                     2-way match   
1595920          Limit                     2-way match   
1595921          Limi

Посмотрим на названия столбцов

In [4]:
log_csv.columns

Index(['User', 'org:resource', 'concept:name', 'Cumulative net worth (EUR)',
       'time:timestamp', 'case:Spend area text', 'case:Company',
       'case:Document Type', 'case:Sub spend area text',
       'case:Purchasing Document', 'case:Purch. Doc. Category name',
       'case:Vendor', 'case:Item Type', 'case:Item Category',
       'case:Spend classification text', 'case:Source', 'case:Name',
       'case:GR-Based Inv. Verif.', 'case:Item', 'case:concept:name',
       'case:Goods Receipt'],
      dtype='object')

Посмотрим на уникальные значения в интересующих нас столбцах, чтобы понять, если ли пропущенные данные

In [5]:
log_csv["org:resource"].value_counts()

NONE        399090
user_002    166353
user_029     71539
user_020     39770
batch_06     38100
             ...  
user_540         1
user_542         1
user_553         1
user_560         1
user_606         1
Name: org:resource, Length: 628, dtype: int64

In [6]:
log_csv["concept:name"].value_counts()

Record Goods Receipt                   314097
Create Purchase Order Item             251734
Record Invoice Receipt                 228760
Vendor creates invoice                 219919
Clear Invoice                          194393
Record Service Entry Sheet             164975
Remove Payment Block                    57136
Create Purchase Requisition Item        46592
Receive Order Confirmation              32065
Change Quantity                         21449
Change Price                            12423
Delete Purchase Order Item               8875
Change Approval for Purchase Order       7541
Cancel Invoice Receipt                   7096
Vendor creates debit memo                6255
Change Delivery Indicator                3289
Cancel Goods Receipt                     3096
SRM: In Transfer to Execution Syst.      1765
SRM: Complete                            1628
SRM: Created                             1628
SRM: Awaiting Approval                   1628
SRM: Document Completed           

In [7]:
log_csv["case:concept:name"].value_counts()

4507027474_00010    990
4507000440_00010    923
4507004396_00020    868
4507000647_00010    833
4507001973_00010    745
                   ... 
4508075459_00870      1
4508075459_00880      1
4508075459_00890      1
4508075459_00900      1
4508075455_00430      1
Name: case:concept:name, Length: 251734, dtype: int64

Оставим от всего журнала всего 4 интересующие нас колонки для простоты работы. Идентификатор случая (case id) - это Заказ, Деятельность (activity name)	- это Изменение статуса заказа, Ресурс (resource) - Исполнитель изменения (система или оператор),
Отметка времени (timestamp)	- Дата и время регистрации изменения.


In [8]:
log_csv_reduced = log_csv[[
                            "case:concept:name",
                            "concept:name",
                            "org:resource", 
                            "time:timestamp",
                        ]]
log_csv_reduced

case:concept:name                         concept:name org:resource  \
0        2000000000_00001                         SRM: Created     batch_00   
1        2000000000_00001                        SRM: Complete     batch_00   
2        2000000000_00001               SRM: Awaiting Approval     batch_00   
3        2000000000_00001              SRM: Document Completed     batch_00   
4        2000000000_00001  SRM: In Transfer to Execution Syst.     batch_00   
...                   ...                                  ...          ...   
1595918  4508076348_00090   Change Approval for Purchase Order     user_603   
1595919  4508076348_00100           Create Purchase Order Item     user_602   
1595920  4508076348_00100   Change Approval for Purchase Order     user_603   
1595921  4508076348_00110           Create Purchase Order Item     user_602   
1595922  4508076348_00110   Change Approval for Purchase Order     user_603   

                   time:timestamp  
0       2018-01-02 12:53:00+00:00  
1       2018-01-02 13:53:00+00:00  
2       2018-01-02 13:53:00+00:00  
3       2018-01-02 13:53:00+00:00  
4       2018-01-02 13:53:00+00:00  
...                           ...  
1595918 2019-01-17 14:00:00+00:00  
1595919 2019-01-17 13:32:00+00:00  
1595920 2019-01-17 14:00:00+00:00  
1595921 2019-01-17 13:32:00+00:00  
1595922 2019-01-17 14:00:00+00:00  

[1595923 rows x 4 columns]

Заменим NONE у исполнителей на Unregistered для большей информативности

In [9]:
log_csv_reduced["org:resource"] = log_csv_reduced["org:resource"].replace(
    "NONE", 
    "Unregistered",
)

C:\Users\barse\AppData\Local\Temp\ipykernel_23692\4222081968.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  log_csv_reduced["org:resource"] = log_csv_reduced["org:resource"].replace(


Посмотрим, сколько всего существует различных путей исполнения заказа  

In [10]:
from pm4py.algo.filtering.log.variants import variants_filter
from pm4py.statistics.traces.generic.log import case_statistics

variants = variants_filter.get_variants(log)
print(f"Всего различных путей: {len(variants)}")

Всего различных путей: 11973


In [11]:
variants_count = case_statistics.get_variant_statistics(log_csv_reduced)

Исследуем, какие наиболее частые пути исполнения заказа были выявлены из журнала

In [12]:
variants_count[:10]

[{'variant': ('Create Purchase Order Item',
   'Vendor creates invoice',
   'Record Goods Receipt',
   'Record Invoice Receipt',
   'Clear Invoice'),
  'count': 50286},
 {'variant': ('Create Purchase Order Item',
   'Record Goods Receipt',
   'Vendor creates invoice',
   'Record Invoice Receipt',
   'Clear Invoice'),
  'count': 30798},
 {'variant': ('Create Purchase Order Item', 'Record Goods Receipt'),
  'count': 12214},
 {'variant': ('Create Purchase Order Item',
   'Vendor creates invoice',
   'Record Goods Receipt',
   'Record Invoice Receipt',
   'Remove Payment Block',
   'Clear Invoice'),
  'count': 11383},
 {'variant': ('Create Purchase Order Item',
   'Receive Order Confirmation',
   'Record Goods Receipt',
   'Vendor creates invoice',
   'Record Invoice Receipt',
   'Clear Invoice'),
  'count': 9694},
 {'variant': ('Create Purchase Requisition Item',
   'Create Purchase Order Item',
   'Vendor creates invoice',
   'Record Goods Receipt',
   'Record Invoice Receipt',
   'Clear

In [13]:
import pandas as pd

log_csv_reduced["time:timestamp"] = pd.to_datetime(
                                        log_csv_reduced["time:timestamp"])

C:\Users\barse\AppData\Local\Temp\ipykernel_23692\3753015368.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  log_csv_reduced["time:timestamp"] = pd.to_datetime(


In [14]:
log_csv_reduced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1595923 entries, 0 to 1595922
Data columns (total 4 columns):
 #   Column             Non-Null Count    Dtype              
---  ------             --------------    -----              
 0   case:concept:name  1595923 non-null  object             
 1   concept:name       1595923 non-null  object             
 2   org:resource       1595923 non-null  object             
 3   time:timestamp     1595923 non-null  datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), object(3)
memory usage: 48.7+ MB


Посмотрим время старта и конца исполнения заказа

In [15]:
case_starts_ends = log_csv_reduced.pivot_table(
    index='case:concept:name', 
    aggfunc={"time:timestamp": ["min", "max"]},
)

case_starts_ends

time:timestamp                          
                                        max                       min
case:concept:name                                                    
2000000000_00001  2018-03-29 13:06:00+00:00 2018-01-02 12:53:00+00:00
2000000001_00001  2019-01-17 10:58:00+00:00 2018-01-03 08:49:00+00:00
2000000002_00001  2019-01-17 10:59:00+00:00 2018-01-04 13:17:00+00:00
2000000003_00001  2018-03-08 12:21:00+00:00 2018-01-08 06:21:00+00:00
2000000003_00002  2018-03-08 12:21:00+00:00 2018-01-08 06:21:00+00:00
...                                     ...                       ...
4508076348_00070  2019-01-17 14:00:00+00:00 2019-01-17 13:32:00+00:00
4508076348_00080  2019-01-17 14:00:00+00:00 2019-01-17 13:32:00+00:00
4508076348_00090  2019-01-17 14:00:00+00:00 2019-01-17 13:32:00+00:00
4508076348_00100  2019-01-17 14:00:00+00:00 2019-01-17 13:32:00+00:00
4508076348_00110  2019-01-17 14:00:00+00:00 2019-01-17 13:32:00+00:00

[251734 rows x 2 columns]

Построим таблицы со вспомогательными данными

In [16]:
case_starts_ends = case_starts_ends.reset_index()
case_starts_ends.columns = ["case:concept:name", "caseend", "casestart"]
log_csv_reduced = log_csv_reduced.merge(
    case_starts_ends, 
    on="case:concept:name",
)
log_csv_reduced["relativetime"] = log_csv_reduced["time:timestamp"] - log_csv_reduced["casestart"]
log_csv_reduced["relativetime_s"] = \
    log_csv_reduced["relativetime"].dt.seconds + \
    86400 * log_csv_reduced["relativetime"].dt.days
log_csv_reduced["relativedays"] = log_csv_reduced["relativetime"].dt.days

log_csv_reduced

case:concept:name                         concept:name org:resource  \
0        2000000000_00001                         SRM: Created     batch_00   
1        2000000000_00001                        SRM: Complete     batch_00   
2        2000000000_00001               SRM: Awaiting Approval     batch_00   
3        2000000000_00001              SRM: Document Completed     batch_00   
4        2000000000_00001  SRM: In Transfer to Execution Syst.     batch_00   
...                   ...                                  ...          ...   
1595918  4508076348_00090   Change Approval for Purchase Order     user_603   
1595919  4508076348_00100           Create Purchase Order Item     user_602   
1595920  4508076348_00100   Change Approval for Purchase Order     user_603   
1595921  4508076348_00110           Create Purchase Order Item     user_602   
1595922  4508076348_00110   Change Approval for Purchase Order     user_603   

                   time:timestamp                   caseend  \
0       2018-01-02 12:53:00+00:00 2018-03-29 13:06:00+00:00   
1       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
2       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
3       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
4       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
...                           ...                       ...   
1595918 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595919 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595920 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595921 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595922 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   

                        casestart    relativetime  relativetime_s  \
0       2018-01-02 12:53:00+00:00 0 days 00:00:00               0   
1       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
2       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
3       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
4       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
...                           ...             ...             ...   
1595918 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595919 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595920 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595921 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595922 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   

         relativedays  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  
...               ...  
1595918             0  
1595919             0  
1595920             0  
1595921             0  
1595922             0  

[1595923 rows x 9 columns]

In [17]:
log_csv_reduced['caselength'] = log_csv_reduced['caseend'] - \
                                log_csv_reduced['casestart']

In [18]:
log_csv_reduced

case:concept:name                         concept:name org:resource  \
0        2000000000_00001                         SRM: Created     batch_00   
1        2000000000_00001                        SRM: Complete     batch_00   
2        2000000000_00001               SRM: Awaiting Approval     batch_00   
3        2000000000_00001              SRM: Document Completed     batch_00   
4        2000000000_00001  SRM: In Transfer to Execution Syst.     batch_00   
...                   ...                                  ...          ...   
1595918  4508076348_00090   Change Approval for Purchase Order     user_603   
1595919  4508076348_00100           Create Purchase Order Item     user_602   
1595920  4508076348_00100   Change Approval for Purchase Order     user_603   
1595921  4508076348_00110           Create Purchase Order Item     user_602   
1595922  4508076348_00110   Change Approval for Purchase Order     user_603   

                   time:timestamp                   caseend  \
0       2018-01-02 12:53:00+00:00 2018-03-29 13:06:00+00:00   
1       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
2       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
3       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
4       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
...                           ...                       ...   
1595918 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595919 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595920 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595921 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595922 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   

                        casestart    relativetime  relativetime_s  \
0       2018-01-02 12:53:00+00:00 0 days 00:00:00               0   
1       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
2       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
3       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
4       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
...                           ...             ...             ...   
1595918 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595919 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595920 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595921 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595922 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   

         relativedays       caselength  
0                   0 86 days 00:13:00  
1                   0 86 days 00:13:00  
2                   0 86 days 00:13:00  
3                   0 86 days 00:13:00  
4                   0 86 days 00:13:00  
...               ...              ...  
1595918             0  0 days 00:28:00  
1595919             0  0 days 00:28:00  
1595920             0  0 days 00:28:00  
1595921             0  0 days 00:28:00  
1595922             0  0 days 00:28:00  

[1595923 rows x 10 columns]

Посмотрим, какие процессы прошел (или не прошел) каждый заказ

In [19]:
document_events = pd.crosstab(
    log_csv_reduced["case:concept:name"], 
    log_csv_reduced["concept:name"],
)

document_events

concept:name       Block Purchase Order Item  Cancel Goods Receipt  \
case:concept:name                                                    
2000000000_00001                           0                     0   
2000000001_00001                           0                     0   
2000000002_00001                           0                     1   
2000000003_00001                           0                     0   
2000000003_00002                           0                     0   
...                                      ...                   ...   
4508076348_00070                           0                     0   
4508076348_00080                           0                     0   
4508076348_00090                           0                     0   
4508076348_00100                           0                     0   
4508076348_00110                           0                     0   

concept:name       Cancel Invoice Receipt  Cancel Subsequent Invoice  \
case:concept:name                                                      
2000000000_00001                        0                          0   
2000000001_00001                        0                          0   
2000000002_00001                        0                          0   
2000000003_00001                        0                          0   
2000000003_00002                        0                          0   
...                                   ...                        ...   
4508076348_00070                        0                          0   
4508076348_00080                        0                          0   
4508076348_00090                        0                          0   
4508076348_00100                        0                          0   
4508076348_00110                        0                          0   

concept:name       Change Approval for Purchase Order  Change Currency  \
case:concept:name                                                        
2000000000_00001                                    0                0   
2000000001_00001                                    0                0   
2000000002_00001                                    0                0   
2000000003_00001                                    0                0   
2000000003_00002                                    0                0   
...                                               ...              ...   
4508076348_00070                                    1                0   
4508076348_00080                                    1                0   
4508076348_00090                                    1                0   
4508076348_00100                                    1                0   
4508076348_00110                                    1                0   

concept:name       Change Delivery Indicator  Change Final Invoice Indicator  \
case:concept:name                                                              
2000000000_00001                           0                               0   
2000000001_00001                           0                               0   
2000000002_00001                           0                               0   
2000000003_00001                           0                               0   
2000000003_00002                           0                               0   
...                                      ...                             ...   
4508076348_00070                           0                               0   
4508076348_00080                           0                               0   
4508076348_00090                           0                               0   
4508076348_00100                           0                               0   
4508076348_00110                           0                               0   

concept:name       Change Price  Change Quantity  ...  SRM: Held  \
case:concept:name                                 ...              
2000000000_0

Посмотрим на самый "популярный" статус заказа в журнале

In [20]:
document_events.sum()

concept:name
Block Purchase Order Item                 514
Cancel Goods Receipt                     3096
Cancel Invoice Receipt                   7096
Cancel Subsequent Invoice                 491
Change Approval for Purchase Order       7541
Change Currency                            35
Change Delivery Indicator                3289
Change Final Invoice Indicator             11
Change Price                            12423
Change Quantity                         21449
Change Rejection Indicator                  2
Change Storage Location                   415
Change payment term                         7
Clear Invoice                          194393
Create Purchase Order Item             251734
Create Purchase Requisition Item        46592
Delete Purchase Order Item               8875
Reactivate Purchase Order Item            543
Receive Order Confirmation              32065
Record Goods Receipt                   314097
Record Invoice Receipt                 228760
Record Service Entry 

Самый частый статус заказа – Регистрация прихода товара (Record Goods Receipt). Данный статус повторялся 314 097 раз в журнале

Посмотрим, через какое максимальное количество статусов прошел один заказ

In [21]:
numbers = document_events.sum(axis=1).sort_values()

numbers[-10:]

case:concept:name
2000000100_00001    464
4507000449_00060    500
4507013227_00001    509
4507000436_00010    513
4507026474_00001    561
4507001973_00010    745
4507000647_00010    833
4507004396_00020    868
4507000440_00010    923
4507027474_00010    990
dtype: int64

Видим, что максимальное количество процессов, которое прошел один из документов равно 990

In [22]:
log_csv_reduced[log_csv_reduced["concept:name"] == "Clear Invoice"].sort_values(by=["relativetime_s"])

case:concept:name   concept:name org:resource  \
1011021  4508046096_00040  Clear Invoice     user_013   
1011011  4508046096_00030  Clear Invoice     user_013   
1011001  4508046096_00020  Clear Invoice     user_013   
1010991  4508046096_00010  Clear Invoice     user_013   
1594407  4508076225_00020  Clear Invoice     user_604   
...                   ...            ...          ...   
118192   4507004931_00050  Clear Invoice     user_002   
118182   4507004931_00040  Clear Invoice     user_002   
118172   4507004931_00030  Clear Invoice     user_002   
118162   4507004931_00020  Clear Invoice     user_002   
118152   4507004931_00010  Clear Invoice     user_002   

                   time:timestamp                   caseend  \
1011021 2018-08-01 11:12:00+00:00 2018-08-30 11:43:00+00:00   
1011011 2018-08-01 11:12:00+00:00 2018-08-30 11:43:00+00:00   
1011001 2018-08-01 11:12:00+00:00 2018-08-30 11:43:00+00:00   
1010991 2018-08-01 11:12:00+00:00 2018-08-30 11:43:00+00:00   
1594407 2018-11-22 08:43:00+00:00 2018-11-22 08:43:00+00:00   
...                           ...                       ...   
118192  2018-05-09 12:08:00+00:00 2018-05-09 12:08:00+00:00   
118182  2018-05-09 12:08:00+00:00 2018-05-09 12:08:00+00:00   
118172  2018-05-09 12:08:00+00:00 2018-05-09 12:08:00+00:00   
118162  2018-05-09 12:08:00+00:00 2018-05-09 12:08:00+00:00   
118152  2018-05-09 12:08:00+00:00 2018-05-09 12:08:00+00:00   

                        casestart        relativetime  relativetime_s  \
1011021 2018-07-30 08:26:00+00:00     2 days 02:46:00          182760   
1011011 2018-07-30 08:26:00+00:00     2 days 02:46:00          182760   
1011001 2018-07-30 08:26:00+00:00     2 days 02:46:00          182760   
1010991 2018-07-30 08:26:00+00:00     2 days 02:46:00          182760   
1594407 2018-11-19 22:59:00+00:00     2 days 09:44:00          207840   
...                           ...                 ...             ...   
118192  1948-01-26 22:59:00+00:00 25670 days 13:09:00      2217935340   
118182  1948-01-26 22:59:00+00:00 25670 days 13:09:00      2217935340   
118172  1948-01-26 22:59:00+00:00 25670 days 13:09:00      2217935340   
118162  1948-01-26 22:59:00+00:00 25670 days 13:09:00      2217935340   
118152  1948-01-26 22:59:00+00:00 25670 days 13:09:00      2217935340   

         relativedays          caselength  
1011021             2    31 days 03:17:00  
1011011             2    31 days 03:17:00  
1011001             2    31 days 03:17:00  
1010991             2    31 days 03:17:00  
1594407             2     2 days 09:44:00  
...               ...                 ...  
118192          25670 25670 days 13:09:00  
118182          25670 25670 days 13:09:00  
118172          25670 25670 days 13:09:00  
118162          25670 25670 days 13:09:00  
118152          25670 25670 days 13:09:00  

[194393 rows x 10 columns]

Посмотрим, сколько уникальных исполнителей было задействовано при каждом измененении статуса

In [23]:
log_csv_reduced.groupby(by=["concept:name"]).agg(
    {"org:resource" : pd.Series.nunique},
)

org:resource
concept:name                                     
Block Purchase Order Item                       5
Cancel Goods Receipt                          111
Cancel Invoice Receipt                         21
Cancel Subsequent Invoice                      10
Change Approval for Purchase Order             49
Change Currency                                 7
Change Delivery Indicator                     109
Change Final Invoice Indicator                  2
Change Price                                  134
Change Quantity                               166
Change Rejection Indicator                      1
Change Storage Location                        56
Change payment term                             4
Clear Invoice                                  24
Create Purchase Order Item                    185
Create Purchase Requisition Item              233
Delete Purchase Order Item                    136
Reactivate Purchase Order Item                 51
Receive Order Confirmation                     10
Record Goods Receipt                          256
Record Invoice Receipt                         31
Record Service Entry Sheet                      1
Record Subsequent Invoice                      10
Release Purchase Order                         20
Release Purchase Requisition                   38
Remove Payment Block                           12
SRM: Awaiting Approval                         13
SRM: Change was Transmitted                     1
SRM: Complete                                  13
SRM: Created                                   12
SRM: Deleted                                    1
SRM: Document Completed                        13
SRM: Held                                       1
SRM: In Transfer to Execution Syst.            12
SRM: Incomplete                                 1
SRM: Ordered                                    1
SRM: Transaction Completed                      5
SRM: Transfer Failed (E.Sys.)                   1
Set Payment Block                               7
Update Order Confirmation                       7
Vendor creates debit memo                       1
Vendor creates invoice                          1

In [24]:
from pm4py.algo.filtering.log.attributes import attributes_filter

activities_org = attributes_filter.get_attribute_values(
    log_csv_reduced, 
    "org:resource",
)
activities_org

{'batch_00': 27934,
 'user_000': 3722,
 'Unregistered': 399090,
 'user_001': 34563,
 'user_002': 166353,
 'user_004': 5050,
 'user_005': 14672,
 'user_003': 116,
 'user_006': 17636,
 'user_007': 7853,
 'user_008': 1036,
 'user_011': 698,
 'user_009': 118,
 'user_010': 179,
 'user_012': 32707,
 'user_013': 35069,
 'user_014': 184,
 'user_015': 27941,
 'user_018': 1786,
 'user_016': 122,
 'user_019': 29764,
 'user_017': 25,
 'user_020': 39770,
 'user_021': 36,
 'batch_01': 2496,
 'user_022': 75,
 'user_023': 5137,
 'user_024': 797,
 'user_025': 36,
 'user_026': 30,
 'batch_02': 27933,
 'user_027': 279,
 'user_028': 9,
 'batch_03': 17384,
 'user_029': 71539,
 'user_030': 12761,
 'batch_04': 2480,
 'user_031': 7210,
 'user_032': 668,
 'user_033': 3632,
 'user_034': 8991,
 'user_035': 1401,
 'user_036': 14745,
 'user_037': 1400,
 'user_038': 13316,
 'user_039': 1418,
 'user_040': 17627,
 'user_041': 2019,
 'user_042': 7674,
 'user_043': 2780,
 'user_044': 2395,
 'user_045': 13652,
 'user_04

In [25]:
sorted_activities_org = sorted(activities_org.items(), key=lambda x:x[1])

sorted_activities_org

[('user_055', 1),
 ('user_219', 1),
 ('user_245', 1),
 ('user_282', 1),
 ('user_298', 1),
 ('user_334', 1),
 ('user_339', 1),
 ('user_340', 1),
 ('user_341', 1),
 ('user_343', 1),
 ('user_397', 1),
 ('user_406', 1),
 ('user_408', 1),
 ('user_409', 1),
 ('user_414', 1),
 ('user_415', 1),
 ('user_418', 1),
 ('user_420', 1),
 ('user_422', 1),
 ('user_425', 1),
 ('user_426', 1),
 ('user_443', 1),
 ('user_489', 1),
 ('user_534', 1),
 ('user_540', 1),
 ('user_542', 1),
 ('user_553', 1),
 ('user_560', 1),
 ('user_562', 1),
 ('user_575', 1),
 ('user_582', 1),
 ('user_584', 1),
 ('user_586', 1),
 ('user_587', 1),
 ('user_590', 1),
 ('user_593', 1),
 ('user_594', 1),
 ('user_595', 1),
 ('user_596', 1),
 ('user_599', 1),
 ('user_606', 1),
 ('user_315', 2),
 ('user_354', 2),
 ('user_385', 2),
 ('user_386', 2),
 ('user_405', 2),
 ('user_423', 2),
 ('batch_15', 2),
 ('user_522', 2),
 ('user_526', 2),
 ('user_541', 2),
 ('user_548', 2),
 ('user_583', 2),
 ('user_585', 2),
 ('user_600', 2),
 ('user_13

За весь журнал событий больше всего действий выполнял user_002 – 166 353 действия

In [26]:
activities_concept = attributes_filter.get_attribute_values(
    log_csv_reduced, 
    "concept:name",
)
activities_concept

{'SRM: Created': 1628,
 'SRM: Complete': 1628,
 'SRM: Awaiting Approval': 1628,
 'SRM: Document Completed': 1628,
 'SRM: In Transfer to Execution Syst.': 1765,
 'SRM: Ordered': 1628,
 'SRM: Change was Transmitted': 1440,
 'Create Purchase Order Item': 251734,
 'Vendor creates invoice': 219919,
 'Record Goods Receipt': 314097,
 'Record Invoice Receipt': 228760,
 'Clear Invoice': 194393,
 'Record Service Entry Sheet': 164975,
 'SRM: Transfer Failed (E.Sys.)': 45,
 'Cancel Goods Receipt': 3096,
 'Vendor creates debit memo': 6255,
 'Cancel Invoice Receipt': 7096,
 'Change Delivery Indicator': 3289,
 'Remove Payment Block': 57136,
 'SRM: Deleted': 188,
 'Change Price': 12423,
 'Delete Purchase Order Item': 8875,
 'SRM: Transaction Completed': 8,
 'Change Quantity': 21449,
 'Change Final Invoice Indicator': 11,
 'SRM: Incomplete': 6,
 'SRM: Held': 6,
 'Receive Order Confirmation': 32065,
 'Cancel Subsequent Invoice': 491,
 'Reactivate Purchase Order Item': 543,
 'Update Order Confirmation': 

Самый частый статус заказа – Регистрация прихода товара (Record Goods Receipt). Данный статус повторялся 314 097 раз в журнале.

In [27]:
tmp = log_csv_reduced["case:concept:name"].value_counts().reset_index()
tmp.columns = ["case:concept:name", "NumActions"]
tmp

case:concept:name  NumActions
0       4507027474_00010         990
1       4507000440_00010         923
2       4507004396_00020         868
3       4507000647_00010         833
4       4507001973_00010         745
...                  ...         ...
251729  4508075459_00870           1
251730  4508075459_00880           1
251731  4508075459_00890           1
251732  4508075459_00900           1
251733  4508075455_00430           1

[251734 rows x 2 columns]

In [28]:
log_csv_reduced = pd.merge(log_csv_reduced, tmp, on="case:concept:name")
log_csv_reduced

case:concept:name                         concept:name org:resource  \
0        2000000000_00001                         SRM: Created     batch_00   
1        2000000000_00001                        SRM: Complete     batch_00   
2        2000000000_00001               SRM: Awaiting Approval     batch_00   
3        2000000000_00001              SRM: Document Completed     batch_00   
4        2000000000_00001  SRM: In Transfer to Execution Syst.     batch_00   
...                   ...                                  ...          ...   
1595918  4508076348_00090   Change Approval for Purchase Order     user_603   
1595919  4508076348_00100           Create Purchase Order Item     user_602   
1595920  4508076348_00100   Change Approval for Purchase Order     user_603   
1595921  4508076348_00110           Create Purchase Order Item     user_602   
1595922  4508076348_00110   Change Approval for Purchase Order     user_603   

                   time:timestamp                   caseend  \
0       2018-01-02 12:53:00+00:00 2018-03-29 13:06:00+00:00   
1       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
2       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
3       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
4       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
...                           ...                       ...   
1595918 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595919 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595920 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595921 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595922 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   

                        casestart    relativetime  relativetime_s  \
0       2018-01-02 12:53:00+00:00 0 days 00:00:00               0   
1       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
2       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
3       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
4       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
...                           ...             ...             ...   
1595918 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595919 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595920 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595921 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595922 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   

         relativedays       caselength  NumActions  
0                   0 86 days 00:13:00          12  
1                   0 86 days 00:13:00          12  
2                   0 86 days 00:13:00          12  
3                   0 86 days 00:13:00          12  
4                   0 86 days 00:13:00          12  
...               ...              ...         ...  
1595918             0  0 days 00:28:00           2  
1595919             0  0 days 00:28:00           2  
1595920             0  0 days 00:28:00           2  
1595921             0  0 days 00:28:00           2  
1595922             0  0 days 00:28:00           2  

[1595923 rows x 11 columns]

In [29]:
log_csv_reduced["Process_time"] = log_csv_reduced.groupby("case:concept:name")["time:timestamp"].diff(1).shift(-1)

log_csv_reduced

case:concept:name                         concept:name org:resource  \
0        2000000000_00001                         SRM: Created     batch_00   
1        2000000000_00001                        SRM: Complete     batch_00   
2        2000000000_00001               SRM: Awaiting Approval     batch_00   
3        2000000000_00001              SRM: Document Completed     batch_00   
4        2000000000_00001  SRM: In Transfer to Execution Syst.     batch_00   
...                   ...                                  ...          ...   
1595918  4508076348_00090   Change Approval for Purchase Order     user_603   
1595919  4508076348_00100           Create Purchase Order Item     user_602   
1595920  4508076348_00100   Change Approval for Purchase Order     user_603   
1595921  4508076348_00110           Create Purchase Order Item     user_602   
1595922  4508076348_00110   Change Approval for Purchase Order     user_603   

                   time:timestamp                   caseend  \
0       2018-01-02 12:53:00+00:00 2018-03-29 13:06:00+00:00   
1       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
2       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
3       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
4       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
...                           ...                       ...   
1595918 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595919 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595920 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595921 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595922 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   

                        casestart    relativetime  relativetime_s  \
0       2018-01-02 12:53:00+00:00 0 days 00:00:00               0   
1       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
2       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
3       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
4       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
...                           ...             ...             ...   
1595918 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595919 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595920 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595921 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595922 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   

         relativedays       caselength  NumActions    Process_time  
0                   0 86 days 00:13:00          12 0 days 01:00:00  
1                   0 86 days 00:13:00          12 0 days 00:00:00  
2                   0 86 days 00:13:00          12 0 days 00:00:00  
3                   0 86 days 00:13:00          12 0 days 00:00:00  
4                   0 86 days 00:13:00          12 0 days 00:00:00  
...               ...              ...         ...             ...  
1595918             0  0 days 00:28:00           2             NaT  
1595919             0  0 days 00:28:00           2 0 days 00:28:00  
1595920             0  0 days 00:28:00           2             NaT  
1595921             0  0 days 00:28:00           2 0 days 00:28:00  
1595922             0  0 days 00:28:00           2             NaT  

[1595923 rows x 12 columns]

In [30]:
log_csv_reduced["Process_time_in_sec"] = log_csv_reduced["Process_time"].dt.total_seconds()
log_csv_reduced

case:concept:name                         concept:name org:resource  \
0        2000000000_00001                         SRM: Created     batch_00   
1        2000000000_00001                        SRM: Complete     batch_00   
2        2000000000_00001               SRM: Awaiting Approval     batch_00   
3        2000000000_00001              SRM: Document Completed     batch_00   
4        2000000000_00001  SRM: In Transfer to Execution Syst.     batch_00   
...                   ...                                  ...          ...   
1595918  4508076348_00090   Change Approval for Purchase Order     user_603   
1595919  4508076348_00100           Create Purchase Order Item     user_602   
1595920  4508076348_00100   Change Approval for Purchase Order     user_603   
1595921  4508076348_00110           Create Purchase Order Item     user_602   
1595922  4508076348_00110   Change Approval for Purchase Order     user_603   

                   time:timestamp                   caseend  \
0       2018-01-02 12:53:00+00:00 2018-03-29 13:06:00+00:00   
1       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
2       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
3       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
4       2018-01-02 13:53:00+00:00 2018-03-29 13:06:00+00:00   
...                           ...                       ...   
1595918 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595919 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595920 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   
1595921 2019-01-17 13:32:00+00:00 2019-01-17 14:00:00+00:00   
1595922 2019-01-17 14:00:00+00:00 2019-01-17 14:00:00+00:00   

                        casestart    relativetime  relativetime_s  \
0       2018-01-02 12:53:00+00:00 0 days 00:00:00               0   
1       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
2       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
3       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
4       2018-01-02 12:53:00+00:00 0 days 01:00:00            3600   
...                           ...             ...             ...   
1595918 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595919 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595920 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   
1595921 2019-01-17 13:32:00+00:00 0 days 00:00:00               0   
1595922 2019-01-17 13:32:00+00:00 0 days 00:28:00            1680   

         relativedays       caselength  NumActions    Process_time  \
0                   0 86 days 00:13:00          12 0 days 01:00:00   
1                   0 86 days 00:13:00          12 0 days 00:00:00   
2                   0 86 days 00:13:00          12 0 days 00:00:00   
3                   0 86 days 00:13:00          12 0 days 00:00:00   
4                   0 86 days 00:13:00          12 0 days 00:00:00   
...               ...              ...         ...             ...   
1595918             0  0 days 00:28:00           2             NaT   
1595919             0  0 days 00:28:00           2 0 days 00:28:00   
1595920             0  0 days 00:28:00           2             NaT   
1595921             0  0 days 00:28:00           2 0 days 00:28:00   
1595922             0  0 days 00:28:00           2             NaT   

         Process_time_in_sec  
0                     3600.0  
1                        0.0  
2                        0.0  
3                        0.0  
4                        0.0  
...                      ...  
1595918                  NaN  
1595919               1680.0  
1595920                  NaN  
1595921               1680.0  
1595922                  NaN  

[1595923 rows x 13 columns]

К узким местам можно отнести те изменения статуса заказа, которые долго (по сравнению с остальными процедурами) длятся при расчете на 1 изменение статуса, то есть имеют наибольшую среднюю длительность. К ним можно отнести следующие статусы

In [31]:
long_active = log_csv_reduced.groupby('concept:name').agg({'Process_time': 'mean', 'org:resource': lambda l: ",".join(l.unique())})
long_active.nlargest(3, 'Process_time')

Process_time  \
concept:name                                            
SRM: Transfer Failed (E.Sys.)       103 days 10:59:00   
SRM: Transaction Completed           77 days 22:36:00   
Vendor creates debit memo     41 days 16:59:48.921600   

                                                               org:resource  
concept:name                                                                 
SRM: Transfer Failed (E.Sys.)                                      user_003  
SRM: Transaction Completed     user_022,user_026,user_016,user_009,user_010  
Vendor creates debit memo                                      Unregistered

In [32]:
grouped_process = log_csv_reduced.groupby(["concept:name"]).agg({"Process_time_in_sec": ["sum", "min", "max", "mean"]})

In [33]:
grouped_process[('Process_time_in_sec',  'sum')] = grouped_process[('Process_time_in_sec',  'sum')] / 3600 / 24
grouped_process[('Process_time_in_sec',  'min')] = grouped_process[('Process_time_in_sec',  'min')] / 3600 / 24
grouped_process[('Process_time_in_sec',  'max')] = grouped_process[('Process_time_in_sec',  'max')] / 3600 / 24
grouped_process[('Process_time_in_sec',  'mean')] = grouped_process[('Process_time_in_sec',  'mean')] / 3600 / 24

grouped_process[('Process_time_in_sec',  'sum')] = grouped_process[('Process_time_in_sec',  'sum')].round()
grouped_process[('Process_time_in_sec',  'min')] = grouped_process[('Process_time_in_sec',  'min')].round(1)
grouped_process[('Process_time_in_sec',  'max')] = grouped_process[('Process_time_in_sec',  'max')].round(1)
grouped_process[('Process_time_in_sec',  'mean')] = grouped_process[('Process_time_in_sec',  'mean')].round(1)

grouped_process

Process_time_in_sec                       
                                                    sum    min      max   mean
concept:name                                                                  
Block Purchase Order Item                        1703.0    0.0    156.0    4.9
Cancel Goods Receipt                            63691.0    0.0    349.0   24.0
Cancel Invoice Receipt                          27711.0    0.0    162.2    4.8
Cancel Subsequent Invoice                        5646.0    0.0    106.1   12.1
Change Approval for Purchase Order             142234.0    0.0    315.5   23.0
Change Currency                                   451.0    0.0    122.9   13.3
Change Delivery Indicator                       36193.0    0.0    201.4   14.1
Change Final Invoice Indicator                     99.0    0.0     17.2   12.4
Change Price                                   141866.0    0.0    255.1   11.9
Change Quantity                                176700.0    0.0    306.9    8.4
Change Rejection Indicator                          0.0    0.0      0.2    0.1
Change Storage Location                          3252.0    0.0    283.0    8.2
Change payment term                               143.0    0.8     67.0   23.9
Clear Invoice                                  440795.0    0.0    546.4   33.7
Create Purchase Order Item                    2788080.0    0.0    369.9   11.3
Create Purchase Requisition Item                43812.0    0.0    443.3    0.9
Delete Purchase Order Item                       2376.0    0.0    142.0    3.2
Reactivate Purchase Order Item                   4056.0    0.0    117.7    7.7
Receive Order Confirmation                     263932.0    0.0    182.5    8.3
Record Goods Receipt                          2830706.0    0.0    332.3    9.7
Record Invoice Receipt                        6940385.0    0.0    332.4   33.7
Record Service Entry Sheet                      52614.0    0.0    337.9    0.3
Record Subsequent Invoice                        2830.0    0.0    126.0   17.9
Release Purchase Order                           4405.0    0.0     41.2    2.7
Release Purchase Requisition                       30.0    0.0     26.8    0.1
Remove Payment Block                          1819849.0    0.0    232.4   35.2
SRM: Awaiting Approval                              0.0    0.0      0.0    0.0
SRM: Change was Transmitted                      6000.0    0.0    195.9    4.3
SRM: Complete                                       0.0    0.0      0.0    0.0
SRM: Created                                      105.0    0.0      0.1    0.1
SRM: Deleted                                      198.0    0.0     23.9    1.1
SRM: Document Completed                           223.0    0.0     66.0    0.1
SRM: Held                                          18.0    0.0     12.2    3.0
SRM: In Transfer to Execution Syst.               312.0    0.0     61.4    0.2
SRM: Incomplete                                     0.0    0.0      0.0    0.0
SRM: Ordered                                        1.0    0.0      1.3    0.0
SRM: Transaction Completed                        156.0    0.0    155.9   77.9
SRM: Transfer Failed (E.Sys.)                     103.0  103.5    103.5  103.5
Set Payment Block                                 788.0    0.1     63.2   12.5
Update Order Confirmation                        2335.0    0.0     69.9    9.7
Vendor creates debit memo                      260676.0    0.0  25565.3   41.7
Vendor creates invoice                        1940683.0    0.0   9056.5    8.8

In [34]:
tmp = log_csv_reduced[log_csv_reduced["concept:name"] == "Create Purchase Order Item"].index + 1

In [35]:
mean_time = log_csv_reduced.loc[tmp,'relativetime_s'].mean() / 3600
min_time = log_csv_reduced.loc[tmp,'relativetime_s'].min() / 3600
max_time = log_csv_reduced.loc[tmp,'relativetime_s'].max() / 3600

print(f'Mean time {round(mean_time / 24, 1)}')
print(f'Min time {min_time}')
print(f'Max time {round(max_time / 24, 1)}')

Mean time 13.6
Min time 0.0
Max time 25566.7


Среднее время между созданием заказа и последующим изменением его статуса составляет 13.6 дней

**Построим модели**

1. Inductive miner

In [36]:
from pm4py.algo.discovery.inductive import algorithm as inductive_miner
from pm4py.visualization.process_tree import visualizer as pt_visualizer
from pm4py.convert import convert_to_petri_net as pt_converter

tree = inductive_miner.apply(log)

net, initial_marking, final_marking = pt_converter(tree)

In [37]:
from pm4py.algo.evaluation.simplicity import algorithm as simplicity_evaluator
from pm4py.algo.evaluation.replay_fitness import algorithm as replay_fitness_evaluator
from pm4py.algo.evaluation.precision import algorithm as precision_evaluator
from pm4py.algo.evaluation.generalization import algorithm as generalization_evaluator
from pm4py.visualization.petri_net import visualizer as pn_visualizer

In [38]:
from pm4py.algo.evaluation.simplicity import algorithm as simplicity_evaluator
from pm4py.algo.evaluation.replay_fitness import algorithm as replay_fitness_evaluator
from pm4py.algo.evaluation.precision import algorithm as precision_evaluator
from pm4py.algo.evaluation.generalization import algorithm as generalization_evaluator

simplicity = simplicity_evaluator.apply(net)
print(f"Простота равна: {simplicity}\n")

fitness = replay_fitness_evaluator.apply(
    log, 
    net, 
    initial_marking, 
    final_marking, 
    variant=replay_fitness_evaluator.Variants.TOKEN_BASED,
)
print(f"Пригодность воспроизведения, используя метод на основе токенов, равна {fitness}\n")

prec = precision_evaluator.apply(
    log, 
    net, 
    initial_marking, 
    final_marking, 
    variant=precision_evaluator.Variants.ETCONFORMANCE_TOKEN,
)
print(f"\nТочность с помощью метода ETConformance равна {prec}\n")

gen = generalization_evaluator.apply(
    log, 
    net, initial_marking, 
    final_marking,
)
print(f"\nОбобщение равно {gen}\n")

Простота равна: 0.5840455840455839



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [00:49<00:00, 240.05it/s]


Пригодность воспроизведения, используя метод на основе токенов, равна {'perc_fit_traces': 99.94478298521456, 'average_trace_fitness': 0.9999950895574359, 'log_fitness': 0.9999902873309994, 'percentage_of_fitting_traces': 99.94478298521456}



replaying log with TBR, completed variants :: 100%|██████████| 203425/203425 [1:13:09<00:00, 46.34it/s]  



Точность с помощью метода ETConformance равна 0.23603956995852582



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [00:53<00:00, 224.88it/s]



Обобщение равно 0.9031643185303689



2. Heuristic Miner

In [39]:
from pm4py.algo.discovery.heuristics import algorithm as heuristics_miner
parameters = {pn_visualizer.Variants.FREQUENCY.value.Parameters.FORMAT: "png"}

net1, initial_marking1, final_marking1 = heuristics_miner.apply(log)

In [40]:
simplicity1 = simplicity_evaluator.apply(net1)
print(f"Простота равна: {simplicity1}\n")

fitness1 = replay_fitness_evaluator.apply(
    log, 
    net1, 
    initial_marking1, 
    final_marking1, 
    variant=replay_fitness_evaluator.Variants.TOKEN_BASED,
)
print(f"Пригодность воспроизведения, используя метод на основе токенов, равна {fitness1}\n")

prec1 = precision_evaluator.apply(
    log, 
    net1, 
    initial_marking1, 
    final_marking1, 
    variant=precision_evaluator.Variants.ETCONFORMANCE_TOKEN,
)
print(f"\nТочность с помощью метода ETConformance равна {prec1}\n")

gen1 = generalization_evaluator.apply(
    log, 
    net1, 
    initial_marking1, 
    final_marking1,
)
print(f"\nОбобщение равно {gen1}\n")

Простота равна: 0.4502617801047121



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [01:12<00:00, 164.30it/s]


Пригодность воспроизведения, используя метод на основе токенов, равна {'perc_fit_traces': 2.0080720125211533, 'average_trace_fitness': 0.8291260665901008, 'log_fitness': 0.83674200968121, 'percentage_of_fitting_traces': 2.0080720125211533}



replaying log with TBR, completed variants :: 100%|██████████| 203425/203425 [19:08<00:00, 177.10it/s] 



Точность с помощью метода ETConformance равна 0.839972538335324



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [01:17<00:00, 153.80it/s]



Обобщение равно 0.7701334432994823



3. Alpha Miner

In [41]:
from pm4py.algo.discovery.alpha import algorithm as alpha_miner
parameters = {pn_visualizer.Variants.FREQUENCY.value.Parameters.FORMAT: "png"}

net2, initial_marking2, final_marking2 = alpha_miner.apply(log)

In [42]:
simplicity2 = simplicity_evaluator.apply(net2)
print(f"Простота равна: {simplicity2}\n")

fitness2 = replay_fitness_evaluator.apply(
    log, 
    net2, 
    initial_marking2, 
    final_marking2, 
    variant=replay_fitness_evaluator.Variants.TOKEN_BASED,
)
print(f"Пригодность воспроизведения, используя метод на основе токенов, равна {fitness2}\n")

prec2 = precision_evaluator.apply(
    log, 
    net2, 
    initial_marking2, 
    final_marking2, 
    variant=precision_evaluator.Variants.ETCONFORMANCE_TOKEN,
)
print(f"\nТочность с помощью метода ETConformance равна {prec2}\n")

gen2 = generalization_evaluator.apply(
    log, 
    net2, 
    initial_marking2, 
    final_marking2,
)
print(f"\nОбобщение равно {gen2}\n")

Простота равна: 0.49593495934959353



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [00:15<00:00, 772.46it/s] 


Пригодность воспроизведения, используя метод на основе токенов, равна {'perc_fit_traces': 0.0, 'average_trace_fitness': 0.34094350704844684, 'log_fitness': 0.33904459184056546, 'percentage_of_fitting_traces': 0.0}



replaying log with TBR, completed variants :: 100%|██████████| 203425/203425 [00:50<00:00, 3998.55it/s]



Точность с помощью метода ETConformance равна 0.2208875627719099



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [00:14<00:00, 818.26it/s] 



Обобщение равно 0.9111816998095763



DFG Граф

In [43]:
from pm4py.algo.discovery.dfg import algorithm as dfg_graph
from pm4py.objects.conversion.dfg import converter as dfg_mining
parameters = {pn_visualizer.Variants.FREQUENCY.value.Parameters.FORMAT: "png"}

dfg = dfg_graph.apply(log)
net3, initial_marking3, final_marking3 = dfg_mining.apply(dfg)

In [44]:
simplicity3 = simplicity_evaluator.apply(net3)
print(f"Простота равна: {simplicity3}\n")

fitness3 = replay_fitness_evaluator.apply(
    log, 
    net3, 
    initial_marking3, 
    final_marking3, 
    variant=replay_fitness_evaluator.Variants.TOKEN_BASED,
)
print(f"Пригодность воспроизведения, используя метод на основе токенов, равна {fitness3}\n")

prec3 = precision_evaluator.apply(
    log, 
    net3, 
    initial_marking3, 
    final_marking3, 
    variant=precision_evaluator.Variants.ETCONFORMANCE_TOKEN,
)
print(f"\nТочность с помощью метода ETConformance равна {prec3}\n")

gen3 = generalization_evaluator.apply(
    log, 
    net3, 
    initial_marking3, 
    final_marking3,
)
print(f"\nОбобщение равно {gen3}\n")

Простота равна: 0.3737931034482759



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [01:14<00:00, 159.78it/s]


Пригодность воспроизведения, используя метод на основе токенов, равна {'perc_fit_traces': 0.0, 'average_trace_fitness': 0.6567366326242245, 'log_fitness': 0.72751003027077, 'percentage_of_fitting_traces': 0.0}



replaying log with TBR, completed variants :: 100%|██████████| 203425/203425 [01:42<00:00, 1984.79it/s]



Точность с помощью метода ETConformance равна 1.0



replaying log with TBR, completed variants :: 100%|██████████| 11973/11973 [01:29<00:00, 134.48it/s]



Обобщение равно 0.6597170252190021



Сравним метрики

In [45]:
from tabulate import tabulate

mydata1 = [
    ["inductive miner", simplicity, fitness['average_trace_fitness'], prec, gen], 
    ["heuristics miner", simplicity1, fitness1['average_trace_fitness'], prec1, gen1], 
    ["alpha miner", simplicity2, fitness2['average_trace_fitness'], prec2, gen2], 
    ["DFG", simplicity3, fitness3['average_trace_fitness'], prec3, gen3],
]
head1 = ["Algorithm", "Simplicity", "Fitness", "Precision", "Generaliztion"]
print(tabulate(mydata1, headers=head1, tablefmt="grid"))

+------------------+--------------+-----------+-------------+-----------------+
| Algorithm        |   Simplicity |   Fitness |   Precision |   Generaliztion |
+==================+==============+===========+=============+=================+
| inductive miner  |     0.584046 |  0.999995 |    0.23604  |        0.903164 |
+------------------+--------------+-----------+-------------+-----------------+
| heuristics miner |     0.450262 |  0.829126 |    0.839973 |        0.770133 |
+------------------+--------------+-----------+-------------+-----------------+
| alpha miner      |     0.495935 |  0.340944 |    0.220888 |        0.911182 |
+------------------+--------------+-----------+-------------+-----------------+
| DFG              |     0.373793 |  0.656737 |    1        |        0.659717 |
+------------------+--------------+-----------+-------------+-----------------+


Как видно из полученных метрик, по двум показателям из четырех (Simplicity, Fitness) лидирует Inductive miner, его обобщенность (Generalization) вторая среди оставшихся алгоритмов, однако при этом у него самая низкая точность (Precision). Поэтому, на наш взгляд, он не подходит для выбора в качестве лучшей модели для анализа процесса.

По нашему предположению, лучшим выбором будет Heuristic miner, поскольку ни одна из его метрик не является худшей по сравнению с остальными алгоритмами, а в большинстве случаев она близка к лучшим показателям. Heuristic miner помогает бороться с шумом (например, в нашем журнале событий шумом могут выступать единовременное создание заказа и его удаление). Также данный алгоритм работает с большими журналами (в нашем журнале более полутора миллиона данных).